src/signature.c

```diff
@@ -46,115 +46,301 @@
 
 #warning "TODO pick signing scheme"
 #if defined(SWUPD_LINUX_ROOTFS)
-#define IMPL IMPL_FAKE
+#define IMPL IMPL_REAL
 #endif
 
 #if IMPL != IMPL_FAKE
 
-static X509_STORE *create_store(const char *, const char *, const char *);
 
-static char *VERIF_FAIL = "Signature verification failed";
-static char *XSTORE_FAIL = "XSTORE creation failed";
+bool signing_enabled = false;
+
+static bool validate_signature(FILE *, FILE *);
+static bool validate_certificate(void);
+static int verify_callback(int, X509_STORE_CTX *);
+static bool get_pubkey(const char *);
+static bool get_certificates(void);
 
 static bool initialized = false;
+static EVP_PKEY *pkey;
+static X509 *cert;
+static char *chain;
+static char *crl;
+static char *certificate;
+static char *ca_dirname;
 
-static X509_STORE *x509_store = NULL;
 
 bool signature_initialize(const char *ca_cert_filename)
 {
 	if (initialized) {
 		return true;
 	}
+
 	OpenSSL_add_all_algorithms();
```

Comment:
we can use EVP_add_cipher to add specific ciphers, the algorithm used to encrypt the private key is need for functions like PEM_read_PrivateKey